File: leetcode/p1094/Solution.java
class Solution {
    public boolean carPooling(int[][] trips, int capacity) {
        Arrays.sort(trips, (t0, t1) -> {
            if (t0[1] == t1[1]) {
                return t0[2] - t1[2];
            }
            return t0[1] - t1[1];
        });
        
        Queue<Integer> pq = new PriorityQueue<>();
        int max = 0;
        for (int[] t : trips) {
            while(!pq.isEmpty() && pq.peek() <= t[1]) {
                pq.poll();
            }
            for (int i = 0; i < t[0]; ++i) {
                pq.offer(t[2]);
            }
            max = Math.max(max, pq.size());
        }
        return max <= capacity;
    }
}

File: leetcode/p1094/Solution.kt
class Solution {
    fun carPooling(trips: Array<IntArray>, capacity: Int): Boolean {
        val sorted = trips.sortedBy { it[1] }
        val pq = PriorityQueue<Int> { i0, i1 ->
            sorted[i0][2] - sorted[i1][2]
        }
        var curCapacity = 0
        var max = 0
        for ((i, t) in sorted.withIndex()) {
            while(!pq.isEmpty() && sorted[pq.peek()][2] <= t[1]) {
                curCapacity -= sorted[pq.poll()][0]
            }
            
            pq.offer(i)
            curCapacity += sorted[i][0]
            
            max = Math.max(max, curCapacity);
        }
        return max <= capacity
    }
}
